MinecraftForge/MinecraftForge
# 2nd attempt [1.12.x] Custom Fluids Behave Like Water

**Preface**: This is a 2nd attempt, superseding #4478. After some failed attempts at rebasing (which wiped out my changes to vanilla classes) plus some further changes I wanted to make, I had to start new fork and do a significant re-write. I apologize to those who spent time reviewing the last one, although this PR ultimately is much the same.

**Background**: Custom fluids do not naturally do water-like things such as drown entities, push entities along with flow, water plants, float boats, etc. This PR addresses all these behaviors. The reason it hadn't been done previously is because Material.WATER was hard-coded throughout the vanilla code, so needs a patch for every behavior. Each patch is very simple, but water does a lot of stuff which is why so many files are affected.

**Implementation Approach**:

Material class has several added properties to reflect each waterlike behavior, along with setter and getter. Eg. setCanDrownEntity(). All default to false and existing liquids WATER and LAVA are set appropriately so this is entirely non-breaking.

Almost everywhere there is a check for Material.WATER was called, it is replaced with appropriate material property getter. There are a few exceptions for things related specifically to BlockLiquid as that class explicitly only handles WATER and LAVA (since custom "liquids" should use the fluid system).

The isInsideMaterial() method is replaced with some targeted ForgeHooks versions that pass a predicate instead (for test against the material property). So you can test if you're inside a drownable material, a material that can float a boat, etc. Note that there is one method that tests the entity's head and another that tests the entire bounding box.

**Features Implemented**:

- Flowing fluids can be enabled to push entities like water.
- Fluids can be enabled to drown entities like water. Air supply HUD overlay and DROWN damage work same as water.
- Fluids can be enabled to float boats.
- Fluids can be enabled to water plants. This affects farmland and things like reed generation.
- Fluids can be enabled to spawn water creatures. This also enables fishing.
- Fluids can be enabled to be "swimmable" meaning that any path navigation and AI will treat it as water note type.
- Fluids can be enabled to be absorbed by sponges.
- Fluids can be enabled so FOV changes like under water.
- Fluids can be enabled to emit water splash/bubble particles when moving through them. Note that even if disabled, fishing, drowing and guardian movement will still generate particles. 
- Fluids can be enabled to mix with concrete powder to form concrete blocks.
- Fluids can be enabled to mix with lava to form obsidian.
- Fluids can be enabled to vaporize when certain dimensions support it.
- Fluids will extinguish fire.

**Features Not Implemented (Possible Future Enhancements):**

- Fluids' splash particles are the same as water, might be cool future enhancement to allow custom splash particles, or at least color them to match the fluid color.
- If a sponge is enabled to absorb your fluid, you will get regular water back from the sponge. Would be cool future enhancement for it to remember the fluid.
- If a fluid freezes it turns into standard ice (which if it melts goes to water). Would be cool future enhancement for it to support custom ice blocks.
- All liquids except lava will extinguish fire. I'll probably issue a pull request myself in future to fix this -- need to consider how it would interact with the canBurn() or if it is its own property.
- Only actual water generates dripping sounds. Would be cool future enhancement for other fluids to do that.
- Gases aren't really covered specifically, probably needs its own overhaul (being discussed by others already).
- Swimming movement is the same for all fluids/liquids. The isSwimming() method doesn't change this but rather sets whether the fluid is a valid swim path for the navigator. It would be interesting in the future to change swimming movement based on density and viscosity of the fluid.

**Testing**:

-  Make sure water and lava still works as expected
- Place fluid with universal bucket (Misc creative tab) and try out all the various things you can do (including trying commenting out the property setters in the test mod).

**Summary**:

I know this is a big PR, but the logic is pretty simple and it provides a LOT of functionality with good modder control.
